server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfiler.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.profile.query;
+
+import org.apache.lucene.search.Query;
+import org.opensearch.search.profile.ContextualProfileBreakdown;
+import org.opensearch.search.profile.ProfileResult;
+import org.opensearch.search.profile.Timer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class acts as a thread-local storage for profiling a query with concurrent execution
+ *
+ * @opensearch.internal
+ */
+public final class ConcurrentQueryProfiler extends QueryProfiler {
+
+    private final Map<Long, ConcurrentQueryProfileTree> threadToProfileTree;
+    // The LinkedList does not need to be thread safe, as the map associates thread IDs with LinkedList, and only
+    // one thread will access the LinkedList at a time.
+    private final Map<Long, LinkedList<Timer>> threadToRewriteTimers;
+
+    public ConcurrentQueryProfiler(AbstractQueryProfileTree profileTree) {
+        super(profileTree);
+        long threadId = getCurrentThreadId();
+        // We utilize LinkedHashMap to preserve the insertion order of the profiled queries
+        threadToProfileTree = Collections.synchronizedMap(new LinkedHashMap<>());
+        threadToProfileTree.put(threadId, (ConcurrentQueryProfileTree) profileTree);
+        threadToRewriteTimers = new ConcurrentHashMap<>();
+        threadToRewriteTimers.put(threadId, new LinkedList<>());
+    }
+
+    @Override
+    public ContextualProfileBreakdown<QueryTimingType> getQueryBreakdown(Query query) {
+        ConcurrentQueryProfileTree profileTree = threadToProfileTree.computeIfAbsent(
+            getCurrentThreadId(),
+            k -> new ConcurrentQueryProfileTree()
+        );
+        return profileTree.getProfileBreakdown(query);
+    }
+
+    /**
+     * Removes the last (e.g. most recent) element on ConcurrentQueryProfileTree stack.
+     */
+    @Override
+    public void pollLastElement() {
+        ConcurrentQueryProfileTree concurrentProfileTree = threadToProfileTree.get(getCurrentThreadId());
```

Comment:
Could we be a bit more cautions with that:

```
if (concurrentProfileTree != null) {
    concurrentProfileTree.pollLast();
}
``

Reply:
It's addressed  now. Thank you.